### PR TITLE
Support implementing interface in tests

### PR DESCRIPTION
**Pull request by kokosing**
Support implementing interface in tests

### CODE REVIEW

**Review by losipiuk — APPROVED** [review]

**Comment by kokosing** [comment]
Updated. Previous version failed with:
```
2020-08-21T19:22:06.8997331Z tests               | FATAL: Test class io.prestosql.tempto.csv_table has methods which are public but not explicitly annotated. Are they missing @Test?
2020-08-21T19:22:06.8997642Z tests               | 		public java.util.Set io.prestosql.tempto.internal.convention.ConventionBasedTestProxyGenerator$ConventionBasedTestProxy.getTestGroups()
2020-08-21T19:22:06.9005692Z tests               | 		public java.lang.String io.prestosql.tempto.internal.convention.ConventionBasedTestProxyGenerator$ConventionBasedTestProxy.getTestName()
2020-08-21T19:22:06.9014077Z tests               | 		public io.prestosql.tempto.Requirement io.prestosql.tempto.internal.convention.ConventionBasedTestProxyGenerator$ConventionBasedTestProxy.getRequirements(io.prestosql.tempto.configuration.Configuration)
2020-08-21T19:22:06.9025254Z tests               | 		public void io.prestosql.tempto.internal.convention.ConventionBasedTestProxyGenerator$ConventionBasedTestProxy.test()
2020-08-21T19:22:06.9026580Z tests               | JVM will be terminated
2020-08-21T19:22:08.0365062Z tests               | (exited)
2020-08-21T19:22:08.6001101Z 2020-08-21T19:22:08.578Z	INFO	main	io.prestosql.tests.product.launcher.cli.TestRun	Test container 4c1f0e64e237190f3780f8d9a819969f1adc9300410e78558050c8d232d97c05 is exited, with exitCode 1
2020-08-21T19:22:08.6820131Z 2020-08-21T19:22:08.681Z	INFO	main	io.prestosql.tests.product.launcher.cli.TestRun	Done, cleaning up
2020-08-21T19:22:08.6828567Z 2020-08-21T19:22:08.682Z	INFO	main	io.prestosql.tests.product.launcher.env.Environments	Shutting down previous containers
2020-08-21T19:22:09.1863568Z presto-worker-0     | (exited)
```

File: presto-testng-services/src/main/java/io/prestosql/testng/services/ReportUnannotatedMethods.java [hunk 1]
```diff
@@ -129,4 +130,18 @@ public static boolean isTemptoClass(Class<?> aClass)
     {
         return aClass.getPackage().getName().startsWith("io.prestosql.tempto");
     }
+
+    private static void forceTestngFailure(Runnable runnable)
```
Comment (findepi) [inline presto-testng-services/src/main/java/io/prestosql/testng/services/ReportUnannotatedMethods.java:134]: Please see https://github.com/prestosql/presto/pull/4951 for my take on this.

Comment (wendigo) [reply presto-testng-services/src/main/java/io/prestosql/testng/services/ReportUnannotatedMethods.java:134]: I guess #4951 is more robust as it covers more listeners.

File: presto-testng-services/src/main/java/io/prestosql/testng/services/ReportUnannotatedMethods.java [hunk 2]
```diff
@@ -140,6 +140,6 @@ private static boolean overrides(Method first, Method second)
 
     public static boolean isTemptoClass(Class<?> aClass)
     {
-        return "io.prestosql.tempto".equals(aClass.getPackage().getName());
+        return aClass.getPackage().getName().startsWith("io.prestosql.tempto");
```
Comment (findepi) [inline presto-testng-services/src/main/java/io/prestosql/testng/services/ReportUnannotatedMethods.java]: BTW Maybe we can deprecate & remove this interfaces?

Also, this should not match eg `io.prestosql.temptomat` -- you need to prefix match with a dot at the end.

File: presto-testng-services/src/main/java/io/prestosql/testng/services/ReportUnannotatedMethods.java [hunk 3]
```diff
@@ -77,17 +76,33 @@ private static boolean isTestMethod(Method method)
         if (method.getDeclaringClass() == Object.class) {
             return true;
         }
-        Class<?> superclass = method.getDeclaringClass().getSuperclass();
-        Method overridden;
+        if (method.getDeclaringClass().isInterface()) {
+            return isTemptoClass(method.getDeclaringClass());
```
Comment (findepi) [inline presto-testng-services/src/main/java/io/prestosql/testng/services/ReportUnannotatedMethods.java:81]: you should not exit early if this is an interface, but not tempto related

```suggestion
        if (method.getDeclaringClass().isInterface() && isTemptoClass(method.getDeclaringClass())) {
            return true;
```

File: presto-testng-services/src/main/java/io/prestosql/testng/services/ReportUnannotatedMethods.java [hunk 4]
```diff
@@ -77,17 +76,33 @@ private static boolean isTestMethod(Method method)
...
+        }
+        for (Class<?> interfaceClass : method.getDeclaringClass().getInterfaces()) {
+            if (isTemptoClass(interfaceClass)) {
+                return true;
```
Comment (findepi) [inline presto-testng-services/src/main/java/io/prestosql/testng/services/ReportUnannotatedMethods.java:85]: Does this disable ReportUnannotatedMethods for any class that happens to implement a tempto interface?

File: presto-testng-services/src/main/java/io/prestosql/testng/services/ReportUnannotatedMethods.java [hunk 5]
```diff
@@ -77,17 +76,33 @@ private static boolean isTestMethod(Method method)
...
             // Simplistic override detection
-            overridden = superclass.getMethod(method.getName(), method.getParameterTypes());
+            return Optional.of(base.getMethod(method.getName(), method.getParameterTypes()));
         }
         catch (NoSuchMethodException ignored) {
-            return false;
+            return Optional.empty();
         }
-
-        return isTestMethod(overridden);
```
Comment (findepi) [inline presto-testng-services/src/main/java/io/prestosql/testng/services/ReportUnannotatedMethods.java:90]: You apparently removed recursion from here.

eg what if A.test overrides B.test and B.test overrides C.test and C.test is `@Test` annotated?

**Review by findepi — CHANGES_REQUESTED** [review]

**Comment by findepi** [comment]
@kokosing can you please rebase onto https://github.com/prestosql/presto/pull/4951?

also, if the failure is caued by tempto generated classes, I think we should have some simple
way of telling them apart and ignoring.

would that be possible?

**Review by kokosing — COMMENTED** [review]
@findepi  comments addressed

**Comment by findepi** [comment]
See my take on a stop gap approach -- https://github.com/prestosql/presto/pull/4951/commits/48c6a3322f4252c41e807ee6d2e406dde8275007

if you are ok with this, let's iterate from there

**Comment by kokosing** [comment]
@findepi rebased

**Review by findepi — APPROVED** [review]

**Pull request merged**